session.go

```diff
@@ -4971,9 +4994,147 @@ func (c *Collection) writeOp(op interface{}, ordered bool) (lerr *LastError, err
 	s.m.RLock()
 	safeOp := s.safeOp
 	bypassValidation := s.bypassValidation
+	enableOpMsg := s.experimental["opmsg"]
 	s.m.RUnlock()
 
-	if socket.ServerInfo().MaxWireVersion >= 2 {
+	serverInfo := socket.ServerInfo()
+
+	if enableOpMsg && serverInfo.MaxWireVersion >= 6 {
```

Comment:
Or it could be split into separate methods for 
insertOp, bulkUpdateOp, bulkDeleteOp?
Maybe this way it would be possible to reuse some code. Didn't try it, so can't tell for sure. 
My point is that this code was already not very readable. And it is getting worse. It is not a dig @feliixx of course, we all work with the tools available to us. 
This might impede future merges from other mgo forks though :(. Opinions are welcome.

Reply:
Totally agree with you, this part of the code is quite a mess... I didn't wanted to rewrite the whole thing, and tried to left the old code untouched to avoid introduce bugs. Plus, I wanted to get rid of the old workflow to have something more readable (and avoid some useless locking): 

old way:  
```
  writeOp() -> writeOpCommand() -> run() -> SimpleQuery() -> Query() -> CheckQueryError()
```
new way: 
```
writeOp() -> sendMessage()
```
this result in some code duplication, but I think it's totally worth it!

Reply:
Agreed, modifying this code can be scary. I am fine with some code duplication, and simplifying the flow would be worth it.